DDP-7876: Equation question rendering

alextanlan opened the pull request:
Implemented new Equation (derived) questions rendering.
The question value is received from back-end side according to the related input values (see [DDP-7183 description](https://broadinstitute.atlassian.net/browse/DDP-7183))
The value is read-only and should not be changed/patched.

Screenshots:
![sn2](https://user-images.githubusercontent.com/7396837/166230837-1fbf9e22-7d19-499f-93d6-ba0acba7354f.png)

![sn1](https://user-images.githubusercontent.com/7396837/166230850-64b5abec-82fd-48d5-9fc4-f5912c741d97.png)

alextanlan committed:
Small refactoring

alextanlan committed:
Fix answerValue type for FileAnswer

alextanlan committed:
Add new equation question

alextanlan committed:
Fix/align naming

alextanlan committed:
Add tests

alextanlan commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activitySection.component.ts:
```diff
@@ -66,7 +67,17 @@ export class ActivitySectionComponent implements OnInit, OnDestroy {
+        return block.blockType === type && block.shown;
+    }
+
+    private updateVisibilityAndValidation(visibility: BlockVisibility[]): void {
```
Changed `public` to `private` because it is used in the file only.

alextanlan committed:
Small fix

alextanlan committed:
Add more tests

alextanlan commented on ddp-workspace/projects/ddp-sdk/src/lib/models/activity/answerValue.ts:
```diff
@@ -6,6 +6,7 @@ import { NumericAnswerType } from './numericAnswerType';
 export type AnswerValue =
   | boolean
   | string
+  | string[]
```
Fix for changed earlier` FileAnswer` (not related to the current feature).

josemanuel167 commented on ddp-workspace/projects/ddp-sdk/src/lib/models/activity/activityEquationQuestionBlock.ts:
```diff
@@ -0,0 +1,20 @@
+        return QuestionType.Equation;
+    }
+
+    public hasAnswer(): boolean {
```
Could be safe to remove it'

alextanlan replied:
Sorry, I did not get the point. 
What do you mean, @josemanuel167 ?

`ActivityEquationQuestionBlock` class  should implement inherited abstract `hasAnswer` method 
from its parent abstract class `ActivityQuestionBlock`.

alextanlan committed:
Fix equation answer format

alextanlan committed:
Refactor formatting of equation answer

alextanlan committed:
Update equation questions answers on front-end side

alextanlan committed:
Fix linter warnings

alextanlan committed:
Refactor equations answer format

alextanlan committed:
Fix tests

alextanlan committed:
Merge branch 'develop' into DDP-7876-equation-question

alextanlan committed:
Revert "Refactor equations answer format"

This reverts commit fb99007672290385ac5251395480a0eb4dcda9f3.

alextanlan committed:
Revert "Fix tests"

This reverts commit fca3a2461bbbf62473d1b3fbc4f0b02d39dabdba.

alextanlan committed:
Remove debug stuff (console.log)

alextanlan committed:
Remove unused input properties

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/activityQuestion.component.ts:
```diff
@@ -19,6 +19,7 @@ import { ActivityValidationResult } from '../../../models/activity/activityValid
                 [validationRequested]="validationRequested$ | async"
                 [studyGuid]="studyGuid"
                 [activityGuid]="activityGuid"
+                [triggerChanges]="triggerChanges"
```
Why is this needed?

alextanlan replied:
There was an issue with updating equations inside of a composite question thus I added a hack for trigger updating.
Anyway, it is useless now. I have removed it.

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activitySection.component.ts:
```diff
@@ -66,7 +76,17 @@ export class ActivitySectionComponent implements OnInit, OnDestroy {
+        this.embeddedComponentsValidationStatus.next(reducedValidationStatus);
+    }
+
+    public isCertainTypeOfBlock(block: ActivityBlock, type: BlockType): boolean {
```
Maybe rename to something like shouldBlockBeShown ? (isCertainTypeOfBlock suggests that it is only the type that matters, but block.shown matters too).

alextanlan replied:
Fixed in be820a94cb240af38838e9f6b81bded1d6c6d45f

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activitySection.component.ts:
```diff
@@ -90,44 +110,67 @@ export class ActivitySectionComponent implements OnInit, OnDestroy {
-    public isMailAddress(block: ActivityBlock): block is ActivityInstitutionBlock {
-        return block.blockType === BlockType.MailAddress && block.shown;
+    private updateEquationQuestions(equations: AnswerResponseEquation[]): void {
+        const equationQuestionBlocks = this.section.allChildBlocks()
```
IMO it would be preferable if this component did not have any logic specific to type of the blocks that are contained by it. (I see we had some logic already here regarding activity blocks. Not great/we should revisit that. But this problem for now).
I think the issue is that we need to get the equation results/values down to the correct block/component, right?
This param value `equations` is being extracted from the response via a SubmissionManager subscription.
Perhaps a way to avoid having Equation specific logic in this component would be to have the Equation component or the EquationBlock subscribe the SubmissionManager directly?

alextanlan replied:
My bad, I overcomplicated the logic.
Keeping an equation question tuned with submission answer response is nice idea, thank you!

In addition we should have similar logic in a composite question with some tweaks (we can't update an equation itself inside of a composite. But we should update the composite answer and re-render the composite question).
Nevertheless, it is better than my previous solution.

Fixed in f3a66d7475dd354e4bfa593f67b808a66c9dc60f

alextanlan committed:
Fix naming

alextanlan committed:
Fix after review

  move equations update logic out of activitySection component

alextanlan committed:
Fix tests

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-composite-answer/activityCompositeAnswer.component.ts:
```diff
@@ -42,38 +62,65 @@ export class ActivityCompositeAnswer implements OnChanges {
+                        })
+                    );
+                    this.block.setAnswer(newAnswer, false);
+                    this.childQuestionBlocks = this.rebuildChildQuestions(this.block, newAnswer);
```
Handling the server responses to  update equation here is definitely better than in the section component.
But could the `activityEquationAnswerComponent` itself subscribe to results from submissionmanager, see if its own stableid is referenced in the equation results and update itself if necessary? I don't think any other answer/question component cares, right?

alextanlan replied:
Right, other questions don't care.

alextanlan replied:
However, I can't see a decent way to update an equation only. 
Perhaps, just take a reference to a equation child of the composite question and make a force re-rendering?
Will try.

alextanlan replied:
Fixed in that^ way (https://github.com/broadinstitute/ddp-angular/pull/1222/commits/1dd4e07203e464d261e52c842a1e8371a890319a)

alextanlan committed:
Update equations only

  specially in a composite question

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-equation-answer/activityEquationAnswer.component.ts:
```diff
@@ -0,0 +1,61 @@
+                this.block.setAnswer(equation.values, false);
+            }
+        }
+    }
```
This looks good! A code change suggestion to make this more compact: move the filtering (iteration through equations to find the one that matches this component) to the pipeline in the ngOnInit.

alextanlan replied:
Fixed in f7521bc1cc3665755cf7d381755d896d1804bf4d

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-composite-answer/activityCompositeAnswer.component.ts:
```diff
@@ -42,36 +78,80 @@ export class ActivityCompositeAnswer implements OnChanges {
+                const newValue = [equationsToUpdate[equationAnswerComponent.block.stableId][index]];
+                equationAnswerComponent.block.setAnswer(newValue);
+            });
+    }
```
Are we sure that for composite we need to do all this? The child equation questions are listening for their changes, so this code and the code in the equation component are both updating? Or you just looking in the equation component for answers of size 1? If so maybe we should add a comment?

alextanlan replied:
Removed that, see https://github.com/broadinstitute/ddp-angular/pull/1222#discussion_r866772902

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-composite-answer/activityCompositeAnswer.component.ts:
```diff
@@ -42,36 +78,80 @@ export class ActivityCompositeAnswer implements OnChanges {
+                            } : answer;
+                        })
+                    );
+                    this.block.setAnswer(newAnswer, false);
```
Would be desirable that composite questions don't know about equations either.
This probably can be achieved if each equation.component and/or equation.block knows that it is part of a composite answer AND what its index position it is in.
With this information, by subscribing to the Patch response, the equation component can pick out its own answer ("My stableid is X, my index pos is Y, so this must be my Equation value from the response").

alextanlan replied:
As I wrote [here](https://github.com/broadinstitute/ddp-angular/pull/1222#discussion_r865199735), I just had no clue how to separate single equations vs composite child equations. I wasn't able to receive a reference to the parent (composite) question. And I was not so brave to create such the reference/connection. That was the root of all the mess around. 

Thank you for such an elegant idea with the index position! 
I fixed the logic in this way. And it looks miles better for now.

alextanlan committed:
Small refactoring

alextanlan committed:
Move equation update logic from a composite  answer component to equation itself

alextanlan committed:
Remove unnecessary stuff

 after moving equation update logic from a composite answer component to equation component itself

alextanlan committed:
Revert "Remove unnecessary stuff"

This reverts commit 2bb3561e5418f84e965fcbc7cd99d47d5812bd81.

alextanlan committed:
Remove unnecessary logic of equations update from a composite answer component

alextanlan committed:
Small improvement

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-equation-answer/activityEquationAnswer.component.ts:
```diff
@@ -0,0 +1,57 @@
+        const fullValue = `${DecimalHelper.mapDecimalAnswerToNumber(answer)}`;
+        const [integerPart, decimalPart] = fullValue.split('.');
+        return integerPart + this.getFormattedDecimalPart(decimalPart);
+    }
```
We have similar conversion code elsewhere, e.g.: https://github.com/broadinstitute/ddp-angular/blob/eff7a60134e39a99462d239cb648bb8650d58de1/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activityNumericAnswer.component.ts#L123-L137
Let's consolidate into a utility. Or DecimalAnswer interface have an implementation that include toString() and does the same thing.

alextanlan replied:
Actually, formatting of a `equation question answer` value and formatting of a `decimal question answer` differ a bit 
(even though they have the same type `DecimalAnswer`). 
In case of we don't have any decimal part scale:
 - `equation` should keep the whole decimal part of the answer value (1.23 -> 1.23)
 - `decimal` should cut the whole decimal part (1.23 -> 1)
 
 Moreover, for `decimal question answer` when the component input fires a value update, we should transform just a number, e.g. 5.27 ( not `DecimalAnswer` format).
 
 Anyway, I have raised a separate [PR#1260](https://github.com/broadinstitute/ddp-angular/pull/1260) for that improvement. It is more generic one. However it did not become easier, imo. 
 
 What  are your thoughts?

alextanlan replied:
Merged that PR.

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-composite-answer/activityCompositeAnswer.component.ts:
```diff
@@ -110,19 +133,25 @@ export class ActivityCompositeAnswer implements OnChanges {
-                }));
+                childQuestionBlockRow
+                    // we don't patch an equation question answer because it is read-only
+                    .filter(childQuestionBlock => !(childQuestionBlock instanceof ActivityEquationQuestionBlock))
```
I BELIEVE you can safely remove this line as I don't expect the EquationComponent will emit an answer.

alextanlan replied:
Here we iterate through all `childQuestionBlocks` of the composite question independently of which child question answer was changed. The filtering of equations here is to exclude equation answers from the composite answer before patching. 
Otherwise, the patch of the composite answer fails.

MocanaAtBroad replied:
Ah. OK!This something that was not really specified clearly in the API, but what you are doing is reasonable.
You have been able to verify this against the Basil example?

I still would like to get rid of the `instanceof` as it tends to bind the code to specific implementation of QuestionBlock.
The only kind of block that this code should be aware is either a generic ActivityBlock or QuestionBlock.
Would suggest then that we extend ActivityQuestionBlock to include a method like `generatesAnswers(): boolean` or something like that.
The default value can be true, but for EquationBlock it returns false. Then we can get rid of `instanceof`

alextanlan replied:
1) yep, I verified the behavior on basil dev env (`Equation activity`).
2) so, you mean that check for a `questionBlock.questionType` does not make sense as well, right ? 
3) `generatesAnswers` means that only EquationQuestion does not change its answer because it is read-only. Is that the idea ?

MocanaAtBroad replied:
Yeah. That's the idea. Essentially instead of checking whether a block is instanceof something we can just call a method that answers whether object can or cannot do something. At this time, EquationQuestionBlock (or whatever it is called) would be the only QuestionBlock that would answer the generatesAnswers question with a false.

alextanlan replied:
4.
 > get rid of the instanceof as it tends to bind the code to specific implementation of QuestionBlock.

I see your point here. 
However seems to me, in this particular case **maybe** we have to bind the code to `ActivityEquationQuestionBlock` (as a specific implementation of `QuestionBlock`). Because we have some special requirements for that type of questions. 
Like, don't patch the equation answer (both itself and being a composite child). 
Also we should set position indexes in a composite answer (for Equation only), but it doesn't connect to the fact that it can/can't generate answers. It's just because the Equation has quite "unnormal" behavior, we were not able to align it with other questions behavior on back-end. Thus we need to tweak it on the client/Angular side.

MocanaAtBroad replied:
Yes. Perfectly fine to subclass. It is desirable in this case.
What we want to avoid is have ActivityCompositeAnswer be aware of QuestionBlock specializations or subclasses, like `ActivityEquationQuestionBlock`. Adding the method `generatesAnswers()` to QuestionBlock allows us to not have to check the specific type of the Block. We just ask it whether it can do a thing.

alextanlan replied:
Fixed in e3738dc34bf88d67c2585152262921740f6ab8ae

alextanlan committed:
Improvement of a decimal answer formatting (#1260)

alextanlan committed:
Avoid binding the code to specific implementation of QuestionBlock

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-composite-answer/activityCompositeAnswer.component.ts:
```diff
@@ -176,4 +211,14 @@ export class ActivityCompositeAnswer implements OnChanges {
+        compositeChildQuestionBlocks.forEach((row: ActivityQuestionBlock<any>[], rowIndex: number) => {
+            for (const questionBlock of row) {
+                if (!(questionBlock.generatesAnswers())) {
+                    (questionBlock as ActivityEquationQuestionBlock).compositeRowIndex = rowIndex;
```
OK. You ALMOST got rid of ActivityEquationQuestionBlock!
Just move `compositeRowIndex` up to ActivityQuestionBlock, set it for all the blocks and you are done.

alextanlan replied:
Done

alextanlan committed:
Filter equation answers in composite answer because they are read-only

  On the other hand equations answers are used for display all composite answers as well

alextanlan committed:
Apply compositeRowIndex for any type of questions
